### PR TITLE
sql: remove NormalizableTableReference

### PR DESCRIPTION
**Pull request by RaduBerinde**
The AST uses an intermediary type `NormalizableTableReference`. This holds an
`UnresolvedName` until `Normalize` is called; after which it holds a
`TableName`.

The logic around this type is hard to follow, and its use results in
in-place modification of the AST. It also leads to unnecessary objects (the
TableName has to be a separate object on the heap).

This PR contains a series of changes that together remove this type altogether.
The approach is to call `NormalizeTableName` directly in the parser and store
`TableName`s in nodes. Note that `NormalizeTableName` performs some checks and
puts the information in a different format, but it does not lose information.

We are calling `NormalizeTableName` directly in many places (instead of factoring it into a separate terminal) to avoid having to put `TableName` on the heap unnecessarily.

Note that some commits had to add some temporary hacks that were then cleaned
up in later commits.

@benesch I had to make some cleanup of the table name usage in zone
specifier code (`config/zone.go`), PTAL at the 4th commit ("sql: use TableName
in TableNameWithIndex").

### CODE REVIEW

**Review by knz — APPROVED** [review]
:lgtm: with two nits below

> Reviewed 66 of 66 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31681#-:-LPMvk4-8G6CYHO2dFCp:b-kef45y)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/backupccl/restore.go, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31681#-LPMrdpp7H35RMM60yJr:-LPMrdpqADTpm1GAmnp3:bxcl9uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc58623718c3c103eae66f0aae4359118736ec30/pkg/ccl/backupccl/restore.go#L175)):*
> ```Go
> 				tn.CatalogName = tree.Name(newDB)
> 			}
> 			tn.TableNamePrefix.Format(ctx)
> ```

Can you explain why this needed to change. I think the original format method is OK here?
___
*[pkg/sql/parser/sql.y, line 1166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31681#-LPMts_SAFVkX-60RgB9:-LPMts_SAFVkX-60RgBA:b-4c2mmw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc58623718c3c103eae66f0aae4359118736ec30/pkg/sql/parser/sql.y#L1166)):*
> ```Yacc
> 
> alter_sequence_options_stmt:
>   ALTER SEQUENCE sequence_name sequence_option_list
> ```

It would be nice to call out in the commit message that the normalization is repeated everywhere instead of factored in a few non-terminals (e.g. `table_name` etc) because of concerns for allocations. It's not obvious and the next person to read this will have a WTF reaction to the code duplication.

File: pkg/config/zone.go [hunk 1]
```diff
@@ -105,44 +104,45 @@ func ParseCLIZoneSpecifier(s string) (tree.ZoneSpecifier, error) {
...
+		}
 	}
 	return tree.ZoneSpecifier{
-		TableOrIndex: parsed,
-		Partition:    partition,
+		TableOrIndex: tree.TableNameWithIndex{
+			Table:       tree.MakeTableName(database, table),
+			Index:       parsed.Index,
+			SearchTable: false,
+		},
```
Comment (benesch) [inline pkg/config/zone.go]: This bit LGTM! And hopefully soon we won't need to be parsing these CLI specifiers.

**Review by benesch — APPROVED** [review]

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31681)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/ccl/backupccl/restore.go, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31681#-LPMrdpp7H35RMM60yJr:-LPNf5KmCKl1VoJTVDO5:b-d21tgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc58623718c3c103eae66f0aae4359118736ec30/pkg/ccl/backupccl/restore.go#L175)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you explain why this needed to change. I think the original format method is OK here?
</blockquote></details>

I should have added a comment. If we just call FormatNode, that will just call back into this function in an infinite recursion.. I changed it to create another ctx and use FormatNode.
___
*[pkg/sql/parser/sql.y, line 1166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31681#-LPMts_SAFVkX-60RgB9:-LPNfPxT1ySQC45ZJkla:b-ukducb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc58623718c3c103eae66f0aae4359118736ec30/pkg/sql/parser/sql.y#L1166)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It would be nice to call out in the commit message that the normalization is repeated everywhere instead of factored in a few non-terminals (e.g. `table_name` etc) because of concerns for allocations. It's not obvious and the next person to read this will have a WTF reaction to the code duplication.
</blockquote></details>

I will. I reckon I should squash all these commits before merging, right?

**Review by knz — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31681)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/ccl/backupccl/restore.go, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31681#-LPMrdpp7H35RMM60yJr:-LPPhDWC1O4KPIbJP62R:b11o1if) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc58623718c3c103eae66f0aae4359118736ec30/pkg/ccl/backupccl/restore.go#L175)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I should have added a comment. If we just call FormatNode, that will just call back into this function in an infinite recursion.. I changed it to create another ctx and use FormatNode.
</blockquote></details>

Nice thanks
___
*[pkg/sql/parser/sql.y, line 1166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31681#-LPMts_SAFVkX-60RgB9:-LPPhGXQDEgwWiiuqsMD:bo0k93t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc58623718c3c103eae66f0aae4359118736ec30/pkg/sql/parser/sql.y#L1166)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I will. I reckon I should squash all these commits before merging, right?
</blockquote></details>

haha I hadn't even realized there were multiple commits -- reviewable showed me only one!
yes, please do. I understand this is just 1 unit of work in total.
